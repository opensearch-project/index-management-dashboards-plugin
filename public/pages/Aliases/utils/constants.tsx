/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SortDirection } from "../../../utils/constants";
import { IAlias } from "../interface";

export const DEFAULT_PAGE_SIZE_OPTIONS = [5, 10, 20, 50];

export const DEFAULT_QUERY_PARAMS = {
  from: "0",
  size: "20",
  search: "",
  sortField: "alias" as keyof IAlias,
  sortDirection: SortDirection.DESC,
  status: "",
  dataSourceId: "",
};
