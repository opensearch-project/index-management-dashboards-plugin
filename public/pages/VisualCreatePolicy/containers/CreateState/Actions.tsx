/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { EuiButton, EuiFormRow, EuiDragDropContext, EuiDroppable, EuiSpacer, EuiText, DropResult } from "@elastic/eui";
import DraggableItem from "../../components/DraggableItem";
import EuiFormCustomLabel from "../../components/EuiFormCustomLabel";
import { Action, UIAction } from "../../../../../models/interfaces";

interface ActionsProps {
  actions: Array<UIAction<Action>>;
  onClickDeleteAction: (idx: number) => void;
  onClickEditAction: (action: UIAction<Action>) => void;
  onDragEndActions: (dropResult: DropResult) => void;
  onClickAddAction: () => void;
}

const Actions = ({ actions, onClickDeleteAction, onClickEditAction, onDragEndActions, onClickAddAction }: ActionsProps) => {
  return (
    <>
      <EuiFormCustomLabel title="Actions" helpText="Actions are the operations ISM performs when an index is in a certain state." />

      {!!actions.length && (
        <EuiFormRow fullWidth isInvalid={false} error={null}>
          <EuiDragDropContext onDragEnd={onDragEndActions}>
            <EuiDroppable droppableId="STATE_ACTIONS_DROPPABLE_AREA">
              {actions.map((action, idx) => (
                <DraggableItem
                  key={action.id}
                  content={action.content()}
                  id={action.id}
                  idx={idx}
                  isLast={actions.length - 1 === idx}
                  onClickDelete={() => onClickDeleteAction(idx)}
                  onClickEdit={() => onClickEditAction(action)}
                  draggableType="action"
                />
              ))}
            </EuiDroppable>
          </EuiDragDropContext>
        </EuiFormRow>
      )}

      <EuiSpacer size="s" />

      {!actions.length && (
        <EuiText>
          <p style={{ backgroundColor: "#F5F7FA", padding: "5px" }}>
            <span style={{ color: "grey", fontWeight: 200, fontSize: "15px" }}>No actions have been added.</span>
          </p>
        </EuiText>
      )}

      <EuiSpacer />

      <EuiButton onClick={onClickAddAction}>+ Add action</EuiButton>
    </>
  );
};

export default Actions;
