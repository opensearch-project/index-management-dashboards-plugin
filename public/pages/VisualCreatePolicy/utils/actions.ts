/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { htmlIdGenerator } from "@elastic/eui";
import { RolloverAction, UIAction } from "../../../../models/interfaces";

const makeId = htmlIdGenerator();

export enum ActionType {
  Allocation = "allocation",
  Close = "close",
  Delete = "delete",
  ForceMerge = "force_merge",
  IndexPriority = "index_priority",
  Notification = "notification",
  Open = "open",
  ReadOnly = "read_only",
  ReadWrite = "read_write",
  ReplicaCount = "replica_count",
  Rollover = "rollover",
  Rollup = "rollup",
  Snapshot = "snapshot",
}

export class RolloverUIAction implements UIAction<RolloverAction> {
  id: string;
  action: RolloverAction;
  type = ActionType.Rollover;

  constructor(action: RolloverAction, id: string = makeId()) {
    this.action = action;
    this.id = id;
  }

  content = () => `Rollover`;

  clone = (action: RolloverAction) => new RolloverUIAction(action, this.id);

  render = () => null;

  toAction = () => this.action;
}
