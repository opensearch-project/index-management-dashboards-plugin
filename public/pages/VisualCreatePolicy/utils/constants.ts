/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AliasAction, AllocationAction } from "../../../../models/interfaces";

export enum ActionType {
  Alias = "alias",
  Allocation = "allocation",
  Close = "close",
  Delete = "delete",
  ForceMerge = "force_merge",
  IndexPriority = "index_priority",
  Notification = "notification",
  Open = "open",
  ReadOnly = "read_only",
  ReadWrite = "read_write",
  ReplicaCount = "replica_count",
  Rollover = "rollover",
  Rollup = "rollup",
  Shrink = "shrink",
  Snapshot = "snapshot",
}

export const DEFAULT_POLICY = {
  description: "A simple default policy that changes the replica count between hot and cold states.",
  default_state: "hot",
  states: [
    {
      name: "hot",
      actions: [{ replica_count: { number_of_replicas: 5 } }],
      transitions: [
        {
          state_name: "cold",
          conditions: { min_index_age: "30d" },
        },
      ],
    },
    {
      name: "cold",
      actions: [{ replica_count: { number_of_replicas: 2 } }],
      transitions: [],
    },
  ],
  ism_template: [],
};

export const EMPTY_DEFAULT_POLICY = {
  description: "A sample description of the policy",
  default_state: "",
  states: [],
  ism_template: [],
};

export const DEFAULT_LEGACY_ERROR_NOTIFICATION = {
  destination: {
    slack: {
      url: "<url>",
    },
  },
  message_template: {
    source: "The index {{ctx.index}} failed during policy execution.",
  },
};

export const DEFAULT_ALIAS: AliasAction = {
  alias: {
    actions: [],
  },
};

export const DEFAULT_ALLOCATION: AllocationAction = {
  allocation: {
    require: {},
    include: {},
    exclude: {},
    wait_for: false,
  },
};

export const DEFAULT_CLOSE = {
  close: {},
};
export const DEFAULT_DELETE = {
  delete: {},
};
export const DEFAULT_FORCE_MERGE = {
  force_merge: {
    max_num_segments: 1,
  },
};
export const DEFAULT_INDEX_PRIORITY = {
  index_priority: {
    priority: 1,
  },
};
export const DEFAULT_NOTIFICATION = {
  notification: {
    channel: {
      id: "",
    },
    message_template: {
      source: "",
    },
  },
};
export const DEFAULT_OPEN = {
  open: {},
};
export const DEFAULT_READ_ONLY = {
  read_only: {},
};
export const DEFAULT_READ_WRITE = {
  read_write: {},
};
export const DEFAULT_REPLICA_COUNT = {
  replica_count: {
    number_of_replicas: 1,
  },
};
export const DEFAULT_ROLLOVER = {
  rollover: {
    min_doc_count: 5,
  },
};
export const DEFAULT_ROLLUP = {
  // TODO: In future make actual UI
  rollup: {
    jsonString: JSON.stringify(
      {
        ism_rollup: {
          target_index: "rollup-nyc-taxi-data",
          description: "Example rollup job",
          page_size: 200,
          dimensions: [
            {
              date_histogram: {
                source_field: "tpep_pickup_datetime",
                fixed_interval: "1h",
                timezone: "America/Los_Angeles",
              },
            },
            {
              terms: {
                source_field: "PULocationID",
              },
            },
          ],
          metrics: [
            {
              source_field: "passenger_count",
              metrics: [
                {
                  avg: {},
                },
                {
                  sum: {},
                },
                {
                  max: {},
                },
                {
                  min: {},
                },
                {
                  value_count: {},
                },
              ],
            },
          ],
        },
      },
      null,
      4
    ),
  },
};

export const DEFAULT_SHRINK = {
  shrink: {
    percentage_of_source_shards: 0.5,
    force_unsafe: false,
    target_index_name_template: {
      source: "{{ctx.index}}_shrunken",
    },
  },
  force_unsafe_input: "no",
};

export const DEFAULT_SNAPSHOT = {
  snapshot: {
    repository: "example-repository",
    snapshot: "example-snapshot",
  },
};

export const actions = [
  DEFAULT_ALIAS,
  DEFAULT_ALLOCATION,
  DEFAULT_CLOSE,
  DEFAULT_DELETE,
  DEFAULT_FORCE_MERGE,
  DEFAULT_INDEX_PRIORITY,
  DEFAULT_NOTIFICATION,
  DEFAULT_OPEN,
  DEFAULT_READ_ONLY,
  DEFAULT_READ_WRITE,
  DEFAULT_REPLICA_COUNT,
  DEFAULT_ROLLOVER,
  DEFAULT_ROLLUP,
  DEFAULT_SHRINK,
  DEFAULT_SNAPSHOT,
];

export const ISM_TEMPLATE_INPUT_MAX_WIDTH = "400px";
