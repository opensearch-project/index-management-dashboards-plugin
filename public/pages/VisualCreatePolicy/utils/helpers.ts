/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Action, ISMTemplate, Policy, State, Transition, UIAction } from "../../../../models/interfaces";
import {
  ActionType,
  DEFAULT_ALIAS,
  DEFAULT_ALLOCATION,
  DEFAULT_CLOSE,
  DEFAULT_DELETE,
  DEFAULT_FORCE_MERGE,
  DEFAULT_INDEX_PRIORITY,
  DEFAULT_NOTIFICATION,
  DEFAULT_OPEN,
  DEFAULT_READ_ONLY,
  DEFAULT_READ_WRITE,
  DEFAULT_REPLICA_COUNT,
  DEFAULT_ROLLOVER,
  DEFAULT_ROLLUP,
  DEFAULT_SHRINK,
  DEFAULT_SNAPSHOT,
} from "./constants";
import {
  AliasUIAction,
  AllocationUIAction,
  CloseUIAction,
  DeleteUIAction,
  ForceMergeUIAction,
  IndexPriorityUIAction,
  NotificationUIAction,
  OpenUIAction,
  ReadOnlyUIAction,
  ReadWriteUIAction,
  ReplicaCountUIAction,
  RolloverUIAction,
  RollupUIAction,
  ShrinkUIAction,
  SnapshotUIAction,
} from "../components/UIActions";

export const getConditionContent = (transition: Transition): string => {
  const {
    conditions: {
      min_doc_count: minDocCount = undefined,
      min_index_age: minIndexAge = undefined,
      min_size: minSize = undefined,
      min_rollover_age: minRolloverAge = undefined,
      cron = undefined,
    } = {},
  } = transition;
  if (minSize != undefined) return `Minimum index size is ${minSize}`;
  if (minDocCount != undefined) return `Minimum index doc count is ${minDocCount}`;
  if (minIndexAge != undefined) return `Minimum index age is ${minIndexAge}`;
  if (minRolloverAge != undefined) return `Minimum rollover age is ${minRolloverAge}`;
  if (cron != undefined) return `After cron expression "${cron.cron.expression}" in ${cron.cron.timezone}`;
  return "No condition";
};

export const getUIActionFromData = (action: Action): UIAction<any> => {
  const actionType = Object.keys(action)
    .filter((key) => key !== "timeout" && key !== "retry")
    .pop();
  if (!actionType) throw new Error(`Failed to get action using type [${actionType}]`);
  const uiAction = getUIAction(actionType);
  return uiAction.clone(action);
};

export const getUIAction = (actionType: string): UIAction<any> => {
  switch (actionType) {
    case ActionType.Alias:
      return new AliasUIAction(DEFAULT_ALIAS);
    case ActionType.Allocation:
      return new AllocationUIAction(DEFAULT_ALLOCATION);
    case ActionType.Close:
      return new CloseUIAction(DEFAULT_CLOSE);
    case ActionType.Delete:
      return new DeleteUIAction(DEFAULT_DELETE);
    case ActionType.ForceMerge:
      return new ForceMergeUIAction(DEFAULT_FORCE_MERGE);
    case ActionType.IndexPriority:
      return new IndexPriorityUIAction(DEFAULT_INDEX_PRIORITY);
    case ActionType.Notification:
      return new NotificationUIAction(DEFAULT_NOTIFICATION);
    case ActionType.Open:
      return new OpenUIAction(DEFAULT_OPEN);
    case ActionType.ReadOnly:
      return new ReadOnlyUIAction(DEFAULT_READ_ONLY);
    case ActionType.ReadWrite:
      return new ReadWriteUIAction(DEFAULT_READ_WRITE);
    case ActionType.ReplicaCount:
      return new ReplicaCountUIAction(DEFAULT_REPLICA_COUNT);
    case ActionType.Rollover:
      return new RolloverUIAction(DEFAULT_ROLLOVER);
    case ActionType.Rollup:
      return new RollupUIAction(DEFAULT_ROLLUP);
    case ActionType.Shrink:
      return new ShrinkUIAction(DEFAULT_SHRINK);
    case ActionType.Snapshot:
      return new SnapshotUIAction(DEFAULT_SNAPSHOT);
    default:
      throw new Error(`Action type [${actionType}] not supported`);
  }
};

export const getActionOptions = (actionRepoSingleton: ActionRepository) => {
  return actionRepoSingleton.getAllActionTypes().map((key) => {
    let uiAction;
    let text = key
      .split("_")
      .map((str) => capitalizeFirstLetter(str))
      .join(" ");
    try {
      uiAction = getUIAction(key)?.customDisplayText;
      if (uiAction !== undefined) text = uiAction;
    } catch (e) {}
    return {
      value: key,
      text: text,
    };
  });
};

class ActionRepository {
  repository: { [actionType: string]: [new (action: Action) => UIAction<any>, Action] } = {
    alias: [AliasUIAction, DEFAULT_ALIAS],
    allocation: [AllocationUIAction, DEFAULT_ALLOCATION],
    close: [CloseUIAction, DEFAULT_CLOSE],
    delete: [DeleteUIAction, DEFAULT_DELETE],
    force_merge: [ForceMergeUIAction, DEFAULT_FORCE_MERGE],
    index_priority: [IndexPriorityUIAction, DEFAULT_INDEX_PRIORITY],
    notification: [NotificationUIAction, DEFAULT_NOTIFICATION],
    open: [OpenUIAction, DEFAULT_OPEN],
    read_only: [ReadOnlyUIAction, DEFAULT_READ_ONLY],
    read_write: [ReadWriteUIAction, DEFAULT_READ_WRITE],
    replica_count: [ReplicaCountUIAction, DEFAULT_REPLICA_COUNT],
    rollover: [RolloverUIAction, DEFAULT_ROLLOVER],
    rollup: [RollupUIAction, DEFAULT_ROLLUP],
    shrink: [ShrinkUIAction, DEFAULT_SHRINK],
    snapshot: [SnapshotUIAction, DEFAULT_SNAPSHOT],
  };

  getAllActionTypes = () => {
    return Object.keys(this.repository);
  };

  registerAction = (actionType: string, uiActionCtor: new (action: Action) => UIAction<any>, defaultAction: Action) => {
    if (this.repository.hasOwnProperty(actionType)) {
      throw new Error(`Cannot register an action twice in the repository [type=${actionType}]`);
    }

    this.repository[actionType] = [uiActionCtor, defaultAction];
  };

  getUIActionFromData = (action: Action) => {
    const actionType = Object.keys(action)
      .filter((key) => key !== "timeout" && key !== "retry")
      .pop();
    if (!actionType) throw new Error(`Failed to get action using type [${actionType}]`);
    const uiAction = this.getUIAction(actionType);
    return uiAction.clone(action);
  };

  getUIAction = (actionType: string): UIAction<any> => {
    const uiAction = this.repository[actionType];
    if (!uiAction) throw new Error(`Action type [${actionType}] not supported`);
    return new uiAction[0](uiAction[1]);
  };
}

export const actionRepoSingleton = new ActionRepository();

// Takes in the ismTemplates which could be a single object, array, or not defined and returns them reformatted as a list
export const convertTemplatesToArray = (ismTemplates: ISMTemplate[] | ISMTemplate | null | undefined): ISMTemplate[] => {
  const templates = [];
  // policy.ism_template can be an array of templates or a single template as an object or null
  if (Array.isArray(ismTemplates)) {
    templates.push(...ismTemplates);
  } else if (ismTemplates) {
    templates.push(ismTemplates);
  }
  return templates;
};

export const capitalizeFirstLetter = ([first, ...rest]: string, locale = navigator.language) =>
  first.toLocaleUpperCase(locale) + rest.join("");

export const getOrderInfo = (
  state: State | null,
  states: State[]
): { order: string; afterBeforeState: string; disableOrderSelections: boolean } => {
  const isEditing = !!state;
  const editingStateIdx = states.findIndex((s) => s === state);
  let afterBeforeState = "";
  let order = "after";
  let disableOrderSelections = false;
  if (isEditing) {
    // If there is only one existing state then we are editing the only state and there is nothing to order after/before
    if (states.length === 1) {
      disableOrderSelections = true;
    } else {
      // If the editing state is the first state then we are "Add before $secondState"
      if (editingStateIdx === 0) {
        afterBeforeState = states[editingStateIdx + 1]?.name;
        order = "before";
      } else {
        // otherwise we are "Add after $previousState"
        afterBeforeState = states[editingStateIdx - 1]?.name;
      }
    }
  } else {
    // when creating a new state
    // If there are no existing states then there is nothing to order after/before
    if (!states.length) {
      disableOrderSelections = true;
    } else {
      // Else always order it after the last state
      afterBeforeState = states[states.length - 1]?.name;
    }
  }

  return { order, afterBeforeState, disableOrderSelections };
};

export const getUpdatedStates = (
  state: State,
  editingState: State | null,
  states: State[],
  order: string,
  afterBeforeState: string
): State[] => {
  const someStates: State[] = [];
  const isEditing = !!editingState;
  if ((isEditing && states.length === 1) || (!isEditing && !states.length)) return [state];

  states.forEach((s, idx) => {
    if (s.name === afterBeforeState && order === "before") someStates.push(state);
    if (s.name !== editingState?.name) someStates.push(s);
    if (s.name === afterBeforeState && order === "after") someStates.push(state);
  });

  return someStates;
};

export const getUpdatedPolicy = (
  currentPolicy: Policy,
  updatedState: State,
  editingState: State | null,
  currentStates: State[],
  order: string,
  afterBeforeState: string
): Policy => {
  const updatedStates = getUpdatedStates(updatedState, editingState, currentStates, order, afterBeforeState);
  let defaultState = currentPolicy.default_state;
  // If there is 1 state, change it to this state
  if (updatedStates.length === 1) {
    defaultState = updatedStates[0].name;
  }
  // Change the default state if the state being edited was the default state
  if (editingState && editingState.name === defaultState) {
    // don't bother checking if the name itself changed, just set it regardless to the value from the new state
    defaultState = updatedState.name;
  }

  return {
    ...currentPolicy,
    states: updatedStates,
    default_state: defaultState,
  };
};
