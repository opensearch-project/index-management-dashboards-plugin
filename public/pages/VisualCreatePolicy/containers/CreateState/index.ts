/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import CreateState from "./CreateState";

export default CreateState;
