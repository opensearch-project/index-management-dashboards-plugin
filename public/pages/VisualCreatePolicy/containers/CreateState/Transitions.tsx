/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { EuiButton, EuiFormRow, EuiDragDropContext, EuiDroppable, EuiSpacer, EuiText, DropResult } from "@elastic/eui";
import EuiFormCustomLabel from "../../components/EuiFormCustomLabel";
import DraggableItem from "../../components/DraggableItem";
import { UITransition } from "../../../../../models/interfaces";
import { getConditionContent } from "../../utils/helpers";

interface TransitionsProps {
  transitions: UITransition[];
  onClickDeleteTransition: (idx: number) => void;
  onClickEditTransition: (transition: UITransition) => void;
  onDragEndTransitions: (dropResult: DropResult) => void;
  onClickAddTransition: () => void;
}

const Transitions = ({
  transitions,
  onClickDeleteTransition,
  onClickEditTransition,
  onDragEndTransitions,
  onClickAddTransition,
}: TransitionsProps) => {
  return (
    <>
      <EuiFormCustomLabel
        title="Transitions"
        helpText="Transitions define the conditions that need to be met for a state to change. After all actions in the current state are completed, the policy starts checking the conditions for transitions."
      />
      {!!transitions.length && (
        <EuiFormRow fullWidth isInvalid={false} error={null}>
          <EuiDragDropContext onDragEnd={onDragEndTransitions}>
            <EuiDroppable droppableId="STATE_TRANSITIONS_DROPPABLE_AREA">
              {transitions.map((transition, idx) => (
                <DraggableItem
                  key={transition.id}
                  id={transition.id}
                  content={getConditionContent(transition.transition)}
                  idx={idx}
                  isLast={transitions.length - 1 === idx}
                  onClickDelete={() => onClickDeleteTransition(idx)}
                  onClickEdit={() => onClickEditTransition(transition)}
                  draggableType="transition"
                />
              ))}
            </EuiDroppable>
          </EuiDragDropContext>
        </EuiFormRow>
      )}

      <EuiSpacer size="s" />

      {!transitions.length && (
        <EuiText>
          <p style={{ backgroundColor: "#F5F7FA", padding: "5px" }}>
            <span style={{ color: "grey", fontWeight: 200, fontSize: "15px" }}>No transitions have been added.</span>
          </p>
        </EuiText>
      )}

      <EuiSpacer />

      <EuiButton onClick={onClickAddTransition}>+ Add Transition</EuiButton>
    </>
  );
};

export default Transitions;
