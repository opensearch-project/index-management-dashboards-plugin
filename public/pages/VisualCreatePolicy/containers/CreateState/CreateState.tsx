/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component, ChangeEvent } from "react";
import {
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutFooter,
  EuiFlyoutHeader,
  EuiTitle,
  EuiFlexGroup,
  EuiFlexItem,
  EuiButtonEmpty,
  EuiButton,
  EuiFormRow,
  EuiFieldText,
  EuiHorizontalRule,
  euiDragDropReorder,
  DropResult,
  EuiSelect,
  EuiSpacer,
  EuiText,
  EuiPortal,
} from "@elastic/eui";
import { UIAction, Policy, State, Action, UITransition } from "../../../../../models/interfaces";
import CreateTransition from "../CreateTransition";
import CreateAction from "../CreateAction";
import Actions from "./Actions";
import Transitions from "./Transitions";
import { makeId } from "../../../../utils/helpers";
import { actionRepoSingleton, getOrderInfo } from "../../utils/helpers";

interface CreateStateProps {
  policy: Policy;
  onSaveState: (state: State, states: State[], order: string, afterBeforeState: string) => void;
  onCloseFlyout: () => void;
  state: State | null;
}

interface CreateStateState {
  name: string;
  nameError: string;
  createAction: boolean;
  editAction: UIAction<Action> | null;
  createTransition: boolean;
  editTransition: UITransition | null;
  actions: Array<UIAction<Action>>;
  transitions: UITransition[];
  afterBeforeState: string;
  order: string;
  disableOrderSelections: boolean;
}

export default class CreateState extends Component<CreateStateProps, CreateStateState> {
  constructor(props: CreateStateProps) {
    super(props);

    const { afterBeforeState, order, disableOrderSelections } = getOrderInfo(props.state, props.policy.states);
    this.state = {
      name: props.state?.name || "",
      nameError: "",
      createAction: false,
      editAction: null,
      createTransition: false,
      editTransition: null,
      actions: props.state?.actions?.map((action) => actionRepoSingleton.getUIActionFromData(action)) || [],
      transitions: props.state?.transitions?.map((transition) => ({ id: makeId(), transition })) || [],
      afterBeforeState,
      order,
      disableOrderSelections,
    };
  }

  onChangeStateName = (event: ChangeEvent<HTMLInputElement>) => {
    const { state, policy } = this.props;
    const name = event.target.value;
    const isEditing = !!this.props.state;
    let nameError = "";
    // If we are not editing a state or if we are editing and have changed the name
    if (!isEditing || (isEditing && name !== state?.name)) {
      // then check to make sure a state doesn't already exist in the policy with that name
      if (!!policy.states.find((newState) => newState.name === name)) {
        nameError = "A state with this name already exists.";
      }
    }

    this.setState({ name, nameError });
  };

  onDragEndActions = ({ source, destination }: DropResult) => {
    if (source && destination) {
      const items = euiDragDropReorder(this.state.actions, source.index, destination.index);
      this.setState({ actions: items });
    }
  };

  onDragEndTransitions = ({ source, destination }: DropResult) => {
    if (source && destination) {
      const items = euiDragDropReorder(this.state.transitions, source.index, destination.index);
      this.setState({ transitions: items });
    }
  };

  onClickDeleteAction = (idx: number) => {
    const { actions } = this.state;
    this.setState({ actions: actions.slice(0, idx).concat(actions.slice(idx + 1)) });
  };

  onClickEditAction = (action: UIAction<Action>) => this.setState({ editAction: action });

  onClickAddAction = () => this.setState({ createAction: true });

  onClickDeleteTransition = (idx: number) => {
    const { transitions } = this.state;
    this.setState({ transitions: transitions.slice(0, idx).concat(transitions.slice(idx + 1)) });
  };

  onClickEditTransition = (transition: UITransition) => this.setState({ editTransition: transition });

  onClickAddTransition = () => {
    this.setState({
      createTransition: true,
    });
  };

  onClickSaveTransition = (transition: UITransition) => {
    const { editTransition, transitions } = this.state;
    let newTransitions = [...transitions, transition];
    if (!!editTransition) {
      const foundTransitionIdx = transitions.findIndex(({ id }) => {
        return transition.id === id;
      });

      if (foundTransitionIdx >= 0) {
        newTransitions = transitions
          .slice(0, foundTransitionIdx)
          .concat(transition)
          .concat(transitions.slice(foundTransitionIdx + 1));
      }
    }
    this.setState((state) => ({
      transitions: newTransitions,
      createTransition: false,
      editTransition: null,
    }));
  };

  onClickSaveAction = (action: UIAction<Action>) => {
    const { editAction, actions } = this.state;
    if (action?.action) {
      let newActions = [...actions, action];
      if (!!editAction) {
        // Use edit action id instead of action id.. as current logic of editing an action can end
        // up with a new id if you switch action types, i.e. rollover -> delete will create a new class w/ a new id
        const foundActionIdx = actions.findIndex(({ id }) => editAction.id === id);
        if (foundActionIdx >= 0) {
          newActions = actions
            .slice(0, foundActionIdx)
            .concat(action)
            .concat(actions.slice(foundActionIdx + 1));
        }
      }
      this.setState({
        actions: newActions,
        editAction: null,
        createAction: false,
      });
    }
  };

  onClickCancelAction = () => {
    this.setState({ createAction: false, editAction: null });
  };

  onClickSaveState = () => {
    const { order, afterBeforeState } = this.state;
    const { onSaveState, policy } = this.props;
    onSaveState(
      {
        name: this.state.name,
        actions: this.state.actions.map((action) => action.toAction()),
        transitions: this.state.transitions.map((transition) => transition.transition),
      },
      policy.states,
      order,
      afterBeforeState
    );
  };

  renderDefault = () => {
    const { policy, state } = this.props;
    const { actions, name, nameError, afterBeforeState, order, disableOrderSelections } = this.state;
    // If we are editing a state filter it out from the selectable options
    const stateOptions = policy.states
      .map((newState) => ({ value: newState.name, text: newState.name }))
      .filter((s) => s.value !== state?.name);
    return (
      <>
        <EuiText>
          <h5>State name</h5>
          <span /> {/* Dummy span to get rid of last child styling on h5 */}
        </EuiText>

        <EuiFormRow fullWidth isInvalid={!!nameError} error={nameError}>
          <EuiFieldText
            fullWidth
            isInvalid={!!nameError}
            placeholder="sample_hot_state"
            readOnly={false}
            value={name}
            onChange={this.onChangeStateName}
            data-test-subj="create-state-state-name"
          />
        </EuiFormRow>

        <EuiSpacer />

        <EuiText>
          <h5>Order</h5>
          <span /> {/* Dummy span to get rid of last child styling on h5 */}
        </EuiText>

        <EuiFlexGroup>
          <EuiFlexItem>
            <EuiFormRow>
              <EuiSelect
                disabled={disableOrderSelections}
                options={[
                  { value: "after", text: "Add after" },
                  { value: "before", text: "Add before" },
                ]}
                value={order}
                onChange={(e) => this.setState({ order: e.target.value })}
                aria-label="Retry failed policy from"
              />
            </EuiFormRow>
          </EuiFlexItem>
          <EuiFlexItem>
            <EuiFormRow>
              <EuiSelect
                disabled={disableOrderSelections}
                options={stateOptions}
                value={afterBeforeState}
                onChange={(e) => this.setState({ afterBeforeState: e.target.value })}
                aria-label="Retry failed policy from"
              />
            </EuiFormRow>
          </EuiFlexItem>
        </EuiFlexGroup>

        <EuiHorizontalRule />

        <Actions
          actions={actions}
          onClickDeleteAction={this.onClickDeleteAction}
          onClickEditAction={this.onClickEditAction}
          onDragEndActions={this.onDragEndActions}
          onClickAddAction={this.onClickAddAction}
        />

        <EuiHorizontalRule />

        <Transitions
          transitions={this.state.transitions}
          onDragEndTransitions={this.onDragEndTransitions}
          onClickDeleteTransition={this.onClickDeleteTransition}
          onClickEditTransition={this.onClickEditTransition}
          onClickAddTransition={this.onClickAddTransition}
        />
      </>
    );
  };

  renderDefaultFooter = () => {
    const { onCloseFlyout, state } = this.props;
    const { name, nameError } = this.state;
    const isEditing = !!state;
    return (
      <EuiFlexGroup justifyContent="spaceBetween">
        <EuiFlexItem grow={false}>
          <EuiButtonEmpty iconType="cross" onClick={onCloseFlyout} flush="left">
            Cancel
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton fill disabled={!name.trim().length || !!nameError} onClick={this.onClickSaveState}>
            {isEditing ? "Update state" : "Save state"}
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  render() {
    const { onCloseFlyout, policy, state } = this.props;
    const { name, createAction, editAction, createTransition, editTransition, actions } = this.state;
    const isEditing = !!state;

    let title = `${isEditing ? "Edit" : "Create"} state`;
    if (createTransition || createAction || !!editTransition || !!editAction) title = "State";
    if (name) title += `: ${name}`;

    // Filter out the current editing state if we are editing
    const stateOptions = policy.states.map((s) => s.name);

    let flyoutContent;
    if (createAction || !!editAction)
      flyoutContent = () => (
        <CreateAction
          actions={actions}
          editAction={editAction}
          stateName={name}
          onClickCancelAction={this.onClickCancelAction}
          onClickSaveAction={this.onClickSaveAction}
        />
      );
    if (createTransition || editTransition)
      flyoutContent = () => (
        <CreateTransition
          stateOptions={stateOptions}
          editTransition={editTransition}
          onCloseCreateTransition={() => this.setState({ createTransition: false, editTransition: null })}
          onClickSaveTransition={this.onClickSaveTransition}
        />
      );
    if (!flyoutContent)
      flyoutContent = () => (
        <>
          <EuiFlyoutBody>{this.renderDefault()}</EuiFlyoutBody>
          <EuiFlyoutFooter>{this.renderDefaultFooter()}</EuiFlyoutFooter>
        </>
      );
    return (
      <EuiPortal>
        <EuiFlyout hideCloseButton ownFocus={false} onClose={onCloseFlyout} maxWidth={600} size="m" aria-labelledby="flyoutTitle">
          <EuiFlyoutHeader hasBorder>
            <EuiTitle size="m">
              <h2 id="flyoutTitle">{title}</h2>
            </EuiTitle>
          </EuiFlyoutHeader>
          {flyoutContent()}
        </EuiFlyout>
      </EuiPortal>
    );
  }
}
