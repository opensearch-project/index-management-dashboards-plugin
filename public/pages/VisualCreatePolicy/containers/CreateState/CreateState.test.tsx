/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import CreateState from "./CreateState";
import { DEFAULT_POLICY } from "../../utils/constants";

describe("<CreateState /> spec", () => {
  it("renders the component", () => {
    render(<CreateState policy={DEFAULT_POLICY} onSaveState={() => {}} onCloseFlyout={() => {}} state={null} />);
    expect(document.body.children[1]).toMatchSnapshot();
  });
});
