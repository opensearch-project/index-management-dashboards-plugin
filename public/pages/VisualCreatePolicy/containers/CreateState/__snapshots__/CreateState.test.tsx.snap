// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateState /> spec renders the component 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
